### PR TITLE
feat(backend): added orders crud + Operations creation(including automatic wallet debit+credit)

### PR DESCRIPTION
**Pull request by DipandaAser**
#### This pull request managing orders

- [x] Create an order
- [x] List organisation orders
- [x] List wallet orders
- [x] Update order
- [x] Pay Order
- [x] Process Order
- [x] Delete order

**Why ?**
User need to order items for spent YOTAS 🤑. Admin need to manage those orders by accepting or declining

**How ?**

- Add endpoint listed up here 
- Create function for handle operations(debit, credit, refund)

**Steps to verify:**

- Run migrations with
- Import this [data](https://files.gamesmania.io/storage/ada5b3b0-7ff6-4168-8b6b-3e3a1a53b71f/yotasDB.zip) in database
- Use this [Postman file](https://www.getpostman.com/collections/aa21992de7b4b0376665) or [Run in Postman](https://app.getpostman.com/run-collection/9853418-0ae6918f-d6f3-4128-8589-8e49985d8392?action=collection%2Ffork&collection-url=entityId%3D9853418-0ae6918f-d6f3-4128-8589-8e49985d8392%26entityType%3Dcollection%26workspaceId%3D4446d68e-0938-4fd5-b862-a2d33b14eb7a)

** Screenshots (optional)
yotas

### CODE REVIEW

File: backend/app/organisation/orders_db.go [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+		order.UpdatedAt = time.Now().UTC()
+		err = tx.Save(order).Error
+		if err != nil {
+			return errors.New("can't create order")
+		}
+
+		return nil
+	})
+
+	return err
```
Comment (elhmn) [inline backend/app/organisation/orders_db.go:64]:
```suggestion
	return db.Session.Transaction(func(tx *gorm.DB) error {

		// We first create the order
		order.State = string(orderStateNotPaid)
		order.CreatedAt = time.Now().UTC()
		err := tx.Create(order).Error
		if err != nil {
			return errors.New("can't create order")
		}

		var totalOrderAmount int64
		// we create each order OrdersArticles and summing the order amount
		for _, orderArticle := range orderItems {
			//sum the amount of article
			totalOrderAmount += int64(orderArticle.Quantity) * orderArticle.Article.Price

			orderArticle.OrderID = order.ID
			orderArticle.ArticlePrice = orderArticle.Article.Price
			orderArticle.CreatedAt = time.Now().UTC()
			err = tx.Create(orderArticle.OrdersArticles).Error
			if err != nil {
				return fmt.Errorf("can't create orders process fail when creating order item %s", orderArticle.Article.Name)
			}
		}

		// finally, we update the order amount
		order.TotalAmount = totalOrderAmount
		order.UpdatedAt = time.Now().UTC()
		err = tx.Save(order).Error
		if err != nil {
			return errors.New("can't create order")
		}

		return nil
	})
```

File: backend/app/organisation/orders_handlers.go [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+}
+
+func GetOrderHandler(ctx *gin.Context) {
+
+	/*	organisationID, err := strconv.Atoi(ctx.GetHeader("Tenant"))
+		if err != nil {
+			ctx.String(http.StatusBadRequest, ErrTenantNotProvided.Error())
+			return
+		}*/
+
```
Comment (elhmn) [inline backend/app/organisation/orders_handlers.go:115]: Left over code ?

File: backend/app/organisation/orders_handlers.go [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+	ctx.String(http.StatusOK, "Order processed")
+}
+
+func PayOrderHandler(ctx *gin.Context) {
+
+	/*	organisationID, err := strconv.Atoi(ctx.GetHeader("Tenant"))
+		if err != nil || organisationID == 0 {
+			ctx.String(http.StatusBadRequest, ErrTenantNotProvided.Error())
+			return
+		}*/
```
Comment (elhmn) [inline backend/app/organisation/orders_handlers.go:311]: Left over ?

File: backend/app/organisation/orders_handlers.go [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+
+func ProcessOrderHandler(ctx *gin.Context) {
+
+	/*	organisationID, err := strconv.Atoi(ctx.GetHeader("Tenant"))
+		if err != nil || organisationID == 0 {
+			ctx.String(http.StatusBadRequest, ErrTenantNotProvided.Error())
+			return
+		}*/
+
+	//TODO check if the user is an admin of the organisation
```
Comment (elhmn) [inline backend/app/organisation/orders_handlers.go:258]: Left over

File: backend/app/organisation/orders_handlers.go [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+
+	ctx.JSON(http.StatusOK, order)
+}
+
+func DeleteOrderHandler(ctx *gin.Context) {
+	/*	organisationID, err := strconv.Atoi(ctx.GetHeader("Tenant"))
+		if err != nil || organisationID == 0 {
+			ctx.String(http.StatusBadRequest, ErrTenantNotProvided.Error())
+			return
+		}*/
```
Comment (elhmn) [inline backend/app/organisation/orders_handlers.go:221]: Left over ?

File: backend/app/organisation/orders_handlers.go [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+		article, err := GetOrganisationArticle(order.Items[index].ArticleID, uint(organisationID))
+		if err != nil {
+			ctx.JSON(http.StatusBadRequest, map[string]interface{}{"messages": []string{
+				fmt.Sprintf("The %v item doesn't belong to this organisation", index)}})
+			return
+		}
+
+		order.Items[index].Article.Articles = *article
+		order.Items[index].Article.Pictures, _ = GetArticlePictures(order.Items[index].ArticleID)
+	}
```
Comment (elhmn) [inline backend/app/organisation/orders_handlers.go:205]:
```suggestion
	for _, item := range order.Items {
		article, err := GetOrganisationArticle(item.ArticleID, uint(organisationID))
		if err != nil {
			ctx.JSON(http.StatusBadRequest, map[string]interface{}{"messages": []string{
				fmt.Sprintf("The %v item doesn't belong to this organisation",  item.ArticleID)}})
			return
		}

		item.Article.Articles = *article
		item.Article.Pictures, _ = GetArticlePictures(item.ArticleID)
	}
```

File: backend/app/organisation/orders_handlers.go [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+			return
+		}
+		orderPresenter.Items = append(orderPresenter.Items, &OrderItemPresenter{
+			OrdersArticles: &orderArticles[orderArticleIndex],
+			Article: ArticlesPresenter{
+				Articles: article,
+				Pictures: pictures,
+			},
+		})
+	}
```
Comment (elhmn) [inline backend/app/organisation/orders_handlers.go:167]: Let's try to use the `for _, elem := range elems { ` as much as possible

File: backend/app/organisation/orders_handlers.go [hunk 8]
```diff
@@ -0,0 +1,338 @@
...
+			}
+			ordersPresenter[i].Items = append(ordersPresenter[i].Items, &OrderItemPresenter{
+				OrdersArticles: &orderArticles[orderArticleIndex],
+				Article: ArticlesPresenter{
+					Articles: article,
+					Pictures: pictures,
+				},
+			})
+		}
+	}
```
Comment (elhmn) [inline backend/app/organisation/orders_handlers.go:94]: for range syntax ?

File: backend/app/organisation/orders_handlers.go [hunk 9]
```diff
@@ -0,0 +1,338 @@
...
+
+	offset = (offset - 1) * limit
+	orders, err := GetOrganisationOrders(uint(organisationID), limit, offset)
+	if err != nil {
+		return
+	}
+
+	ordersPresenter := make([]OrderPresenter, len(orders))
+	for i := 0; i < len(orders); i++ {
```
Comment (elhmn) [inline backend/app/organisation/orders_handlers.go:54]: for range syntax

**Review by elhmn — CHANGES_REQUESTED** [review]

File: backend/app/organisation/orders_handlers.go (hunk 2 above)
Comment (DipandaAser) [reply backend/app/organisation/orders_handlers.go:115]: resolved

File: backend/app/organisation/orders_handlers.go (hunk 3 above)
Comment (DipandaAser) [reply backend/app/organisation/orders_handlers.go:311]: resolved

File: backend/app/organisation/orders_handlers.go (hunk 4 above)
Comment (DipandaAser) [reply backend/app/organisation/orders_handlers.go:258]: resolved

File: backend/app/organisation/orders_handlers.go (hunk 5 above)
Comment (DipandaAser) [reply backend/app/organisation/orders_handlers.go:221]: resolved

File: backend/app/organisation/orders_handlers.go (hunk 7 above)
Comment (DipandaAser) [reply backend/app/organisation/orders_handlers.go:167]: resolved

File: backend/app/organisation/orders_handlers.go (hunk 8 above)
Comment (DipandaAser) [reply backend/app/organisation/orders_handlers.go:94]: resolved

File: backend/app/organisation/orders_handlers.go (hunk 9 above)
Comment (DipandaAser) [reply backend/app/organisation/orders_handlers.go:54]: resolved

File: backend/app/organisation/operations_db.go [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+		// create the operation without the operation hash
+		operation := Operations{
+			Model:         db.Model{CreatedAt: time.Now().UTC(), UpdatedAt: time.Now().UTC()},
+			Amount:        amount,
+			WalletId:      walletID,
+			Description:   description,
+			OperationType: string(operationType),
+			Approved:      true,
+			//due to nullity constraint in migrations we give an initial random value to our operation hash
+			OperationHash: fmt.Sprintf("%v", time.Now().UTC()),
```
Comment (Sanix-Darker) [inline backend/app/organisation/operations_db.go:57]: 👍🏾

**Review by Sanix-Darker — APPROVED** [review]
Thank you for this PR !
- Things looks good to me
- Make sure to get latest modifications from the master branch/and fix conflicts
- This PR is for so many subject that for me it was a little bit hard to follow things well, please, if possible in the future, make sure to manage PR one subject at time !

File: backend/app/organisation/operations_db.go (hunk 10 above)
Comment (DipandaAser) [reply backend/app/organisation/operations_db.go:57]: > 
> 
> Thank you for this PR !
> 
>     * Things looks good to me
> 
>     * Make sure to get latest modifications from the master branch/and fix conflicts
> 
>     * This PR is for so many subject that for me it was a little bit hard to follow things well, please, if possible in the future, make sure to manage PR one subject at time !

Thanks. Next time i will manage PR one subject at time

**Comment by elhmn** [comment]
Fantastic job @DipandaAser I am merging this 🥳
